onnxruntime/core/providers/cpu/ml/label_encoder.h

```diff
@@ -53,57 +54,181 @@ class LabelEncoder_2 final : public OpKernel {
-    const Tensor& X = *tensor_pointer;
-    const TensorShape& shape = X.Shape();
-    Tensor& Y = *context->Output(0, shape);
-
-    auto input = X.template DataAsSpan<TKey>();
-    auto output = Y.template MutableDataAsSpan<TValue>();
-
-    for (int64_t i = 0; i < shape.Size(); ++i) {
-      const auto found = _map.find(input[onnxruntime::narrow<size_t>(i)]);
-      if (found == _map.end())
-        output[onnxruntime::narrow<size_t>(i)] = _default_value;
-      else
-        output[onnxruntime::narrow<size_t>(i)] = found->second;
+    const auto* X = context->Input<Tensor>(0);
+    const TensorShape& shape = X->Shape();
+    auto* Y = context->Output(0, shape);
+
+    auto input = X->template DataAsSpan<TKey>();
+    auto output = Y->template MutableDataAsSpan<TValue>();
+    auto input_iter = input.begin();
+    auto output_iter = output.begin();
+    while (input_iter != input.end()) {
+      const auto found = map_.find(*input_iter);
+      *output_iter = found == map_.end() ? default_value_ : found->second;
+      output_iter++;
+      input_iter++;
     }
-
     return Status::OK();
   }
 
  private:
   // Specialize this method to set attribute names. For example, if keys' type
-  // is 64-bit integer, _key_field_name should be "keys_int64s". Field names
+  // is 64-bit integer, key_field_name_ should be "keys_int64s". Field names
   // for other types can be found in ONNX spec.
   void InitializeSomeFields(const OpKernelInfo& info);
 
   // A collection of key-value pairs. Each (a_key, a_value) pair
   // means that the "a_key" in the input would be mapped to "a_value".
-  // If _map doesn't contain "a_key", we use _default_value as its output.
-  InlinedHashMap<TKey, TValue> _map;
-  TValue _default_value;
+  // If map_ doesn't contain "a_key", we use default_value_ as its output.
+  InlinedHashMap<TKey, TValue> map_;
+  TValue default_value_;
   // ONNX attribute name to load keys.
-  std::string _key_field_name;
+  std::string key_field_name_;
   // ONNX attribute name to load values.
-  std::string _value_field_name;
+  std::string value_field_name_;
+};
+
+template <typename T>
+std::vector<T> GetAttribute(const OpKernelInfo& info, const std::string& name, const std::string& tensor_name) {
+  if constexpr (std::is_same_v<T, std::string> || std::is_same_v<T, float> || std::is_same_v<T, int64_t>) {
+    std::vector<T> attrs;
+    if (info.GetAttrs<T>(name, attrs).IsOK()) {
+      return attrs;
+    }
+  }
+  ONNX_NAMESPACE::TensorProto attr_tensor_proto;
+  auto result = info.GetAttr(tensor_name, &attr_tensor_proto);
+  if (name.empty()) {
+    ORT_ENFORCE(result.IsOK(), "LabelEncoder is missing attribute ", tensor_name);
+  } else {
+    ORT_ENFORCE(result.IsOK(), "LabelEncoder is missing attribute ", tensor_name, " or ", name);
+  }
+  size_t tensor_size = 1;
+  for (auto dim : attr_tensor_proto.dims()) {
+    tensor_size *= static_cast<size_t>(dim);
```

Comment:
>tensor_size *= static_cast<size_t>(dim); [](http://example.com/codeflow?start=4&length=40)

We need to compute the result first in int64_t,
then use `narrow<size_t>()` to make sure any data loss is detected.

Reply:
Thanks for the suggestion!